arctic/store/version_store.py

```diff
@@ -148,22 +151,32 @@ def list_symbols(self, all_symbols=False, snapshot=None, regex=None, **kwargs):
             # Match based on user criteria first
             pipeline.append({'$match': query})
         pipeline.extend([
-                        # Id is by insert time which matches version order
-                        {'$sort': bson.SON([('symbol', pymongo.DESCENDING), ('version', pymongo.DESCENDING)])},
-                        # Group by 'symbol'
-                        {'$group': {'_id': '$symbol',
-                                    'deleted': {'$first': '$metadata.deleted'},
-                                    },
-                         },
-                        # Don't include symbols which are part of some snapshot, but really deleted...
-                        {'$match': {'deleted': {'$ne': True}}},
-                        {'$project': {'_id': 0,
-                                      'symbol':  '$_id',
-                                      }
-                         }])
-
-        results = self._versions.aggregate(pipeline)
-        return sorted([x['symbol'] for x in results])
+            # version_custom value is: 2*version + (0 if deleted else 1)
+            # This is used to optimize aggregation query:
+            #  - avoid sorting
+            #  - be able to rely on the latest version (max) for the deleted status
+            #
+            # Be aware of that if you don't use custom sort or if use a sort before $group which utilizes
+            # exactly an existing index, the $group will do best effort to utilize this index:
+            #  - https://jira.mongodb.org/browse/SERVER-4507
+            {'$group': {
+                '_id': '$symbol',
+                'version_custom': {
+                    '$max': {
+                        '$add': [
+                            {'$multiply': ['$version', 2]},
+                            {'$cond': [{'$eq': ['$metadata.deleted', True]}, 1, 0]}
+                        ]
+                    }
+                },
+            }},
+            # Don't include symbols which are part of some snapshot, but really deleted...
+            {'$match': {'version_custom': {'$mod': [2, 0]}}}
+        ])
+
+        # We may hit the group memory limit (100MB), so use allowDiskUse to circumvent this
+        #  - https://docs.mongodb.com/manual/reference/operator/aggregation/group/#group-memory-limit
+        return sorted([x['_id'] for x in self._versions.aggregate(pipeline, allowDiskUse=True)])
```

Comment:
how much latency does this add?

Reply:
We only hit this if the sorting/grouping mem footprint at mongo exceeds 100MB.
With the new pipeline, we don't hit it with our largest libs (100K+ version docs).

I noticed a query which was previously failing with "memory exceeded" OperationFailure, 
took 4sec to complete.
The equivalent query without the sorting part, was ~1.5sec faster.
This totally depends on how heavy is the query.

Reply:
The 'allowDiskUse' is there now as a gueard dog, to avoid failure when/if memory is exceeded.